torch/csrc/jit/passes/onnx/constant_fold.cpp

```diff
@@ -0,0 +1,187 @@
+    auto qe = updated_val.size(0);
+  }
+  return updated_val;
+}
+
+static bool isConstant(Value* val, const ValueToParamPairMap& valsToParamsMap) {
+  auto parentNode = val->node();
+  return (parentNode->kind() == prim::Param && 
+          valsToParamsMap.find(val) != valsToParamsMap.end()) || // Checks val is a parameter and not a real input
+         (parentNode->kind() == onnx::Constant && !parentNode->mustBeNone() &&
+          parentNode->kindOf(attr::value) == AttributeKind::t); // Check other types?
+}
+
+static std::vector<at::Tensor> getValues(Node* node, const ValueToParamPairMap& valsToParamsMap) {
+  size_t numInputs = node->inputs().size();
+  std::vector<at::Tensor> inputTensorValues;
+  inputTensorValues.reserve(numInputs);
+  for (auto val : node->inputs()) {
+    if (val->node()->kind() == prim::Param) {
+      auto itr = valsToParamsMap.find(val);
+      if(itr == valsToParamsMap.end()) {
+        throw std::runtime_error("getValues: Input value not found amongst constant parameters.");
+      }
+      inputTensorValues.push_back(itr->second.second);
+    }
+    else if (val->node()->kind() == onnx::Constant) {
+      inputTensorValues.push_back(val->node()->t(attr::value));
+    }
+    else {
+      throw std::runtime_error("getValues: Unsupported kind of constant node found.");
+    }
+  }
+  AT_ASSERT(inputTensorValues.size() == numInputs);
+  return inputTensorValues;
+}
+
+static void eraseUnusedValuesFromMap(ValueToParamPairMap& valsToParamsMap) {
+  auto it = valsToParamsMap.begin();
+  while (it != valsToParamsMap.end()) {
+    if (!it->first->hasUses()) {
+      it = valsToParamsMap.erase(it);
+    } 
+    else {
+      ++it;
+    }
+  }
+}
+
+// This method updates the block in-place to fold all the one-time 
+// constant-based computations/ops into an initializer node.
+void ConstantFoldONNX(Block* b, ParamMap& paramsDict) {
+  auto sourceNode = b->param_node();
+  AT_ASSERT(sourceNode);
+  auto valsToParamsMap = buildValueToParamsMap(b, paramsDict);
+  // Only the root block is constant-folded. Folding nested blocks is
+  // not supported for now.
+  for (auto it = b->nodes().begin(), end = b->nodes().end(); it != end; ++it) {
+    auto node = *it;
+    if (node->outputs().size() > 1) {
+        // Constant folding for multiple-output nodes not supported. Skip it.
+        continue;
+      }
+    if (!std::all_of(node->inputs().begin(), node->inputs().end(),
+        [&valsToParamsMap](Value* v) { return isConstant(v, valsToParamsMap); })) {
+        // If all the inputs to this node are not either parameter or
+        // onnx::Constant, then skip this node.
+        continue;
+    }
+    auto inputTensorValues = getValues(node, valsToParamsMap);
+    if (inputTensorValues.empty()) {
+        // This is a terminal node with no inputs, such as onnx::Constant. Skip it.
+        continue;
+    }
+    auto updated_val = runTorchBackendForOnnx(node, inputTensorValues);
+    if (updated_val.size(0) == 0) {
```

Comment:
same as above; returning `c10::nullopt` instead of an empty tensor will express your intent more clearly.

Reply:
Done.